cpp/src/arrow/dataset/partition_test.cc

```diff
@@ -1048,5 +1051,60 @@ TEST(TestStripPrefixAndFilename, Basic) {
                                           "year=2019/month=12/day=01"));
 }
 
+TEST_F(TestPartitioning, PartitioningWithDataWithFilePaths) {
+  fs::TimePoint mock_now = std::chrono::system_clock::now();
+  ASSERT_OK_AND_ASSIGN(std::shared_ptr<fs::FileSystem> filesystem,
+                       fs::internal::MockFileSystem::Make(mock_now, {}));
+  auto base_path = "";
+  ASSERT_OK(filesystem->CreateDir(base_path));
+  // Create an Arrow Table
+  auto schema = arrow::schema(
+      {arrow::field("a", arrow::int64()), arrow::field("part", arrow::utf8())});
+  std::vector<std::shared_ptr<arrow::Array>> arrays(2);
+  arrow::NumericBuilder<arrow::Int64Type> builder;
+  ASSERT_OK(builder.AppendValues({0, 1, 2, 3, 4}));
+  ASSERT_OK(builder.Finish(&arrays[0]));
+  arrow::StringBuilder string_builder;
+  std::vector<std::string> partitions = {"experiment/A/f.csv", "experiment/B/f.csv",
+                                         "experiment/A/f.csv", "experiment/C/k.csv",
+                                         "experiment/M/i.csv"};
+  ASSERT_OK(string_builder.AppendValues(partitions));
+  ASSERT_OK(string_builder.Finish(&arrays[1]));
+  auto table = arrow::Table::Make(schema, arrays);
```

Comment:
Can we use `ArrayFromJSON` or even `TableFromJSON` instead of building these step-by-step?

Reply:
Good catch. We should be able to.